---
layout: post
title: Bioinformatics software I frequently use
---

I use some software more frequently in my daily work, and here is a list for them.

# RNA-seq
## Softwares
* [Texudo suite](http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html) is very popular suite for transcriptome alignment, assembling, and quantatitive analysis, composed by [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml), [Tophat](http://tophat.cbcb.umd.edu/manual.html), and [Cufflinks](http://cufflinks.cbcb.umd.edu/).
* STAR
* [MapSplice](http://www.netlab.uky.edu/p/bioinfo/MapSplice2)
* DESeq and DESeq2.
* EdgeR.
* Voom-lima.
## References

# small RNA-seq
## Softwares
* miRDeep.
* miRanalyzer.
* miRBase.
* miRPath.
## References

# ChIP-seq
## Softwares
* HOMER.
* Macs2
* PhantomPeaks.
* ChIP-seq preprocess.
* Bismark.
* diffReps.
* region_analysis.
## References

# Misc
## Softwares
### General purpose
* samtools.
* Picard.
* [bedtools](https://github.com/arq5x/bedtools2).
* ngs.plot.
### QC
* FastQC.

### Motif analysis
* [MEME Suite](http://meme.nbcr.net/meme/intro.html).
### Integration of RNA-seq and ChIP-seq
* GeneOverlaps.
### Pipeline related
* Ruffus.
### Cluster related